### PR TITLE
feat: extract embedded Electron MCP package

### LINKED ISSUE #2: Carry over 13 generic tool factory unit tests from PR A
## Context

Acceptance criterion in agent-labs-dev/nebula-desktop#153 requires the 13 generic tool factory unit tests from PR A to pass in CI. They live in nebula-desktop's \`src/background/mcp/tools/*.test.ts\` and use \`testing.ts\` (vitest fakes for \`McpServer\` / \`BrowserWindow\` / \`webContents\`).

Out of scope of the smoke-test PR (#1). Filed for follow-up.

## Scope

- Add \`vitest\` (or whatever test runner the team picks) as a dev dep
- Copy \`src/testing.ts\` and the 13 \`tools/*.test.ts\` files from nebula-desktop's \`src/background/mcp/\`
- Adapt imports for the new repo layout
- Add a \`pnpm test:unit\` script
- Wire it into the future CI workflow (separate issue)

## Acceptance

- [ ] All 13 tool factory unit tests pass under \`pnpm test:unit\` locally
- [ ] No vitest/test-runner regressions in the smoke test

### LINKED ISSUE #3: Repo docs scaffolding: README, CONTRIBUTING, SECURITY, CODE_OF_CONDUCT, PR/issue templates, dependabot
## Context

Acceptance criterion in agent-labs-dev/nebula-desktop#153. The bootstrap commit landed bare LICENSE/README/CODEOWNERS/dependabot. This issue tracks the rest.

## Scope

- \`README.md\` — lead with embedded-vs-external positioning, then quickstart, API reference, \`recommendedGuards()\` snippet (separate issue), contributing pointer
- \`CONTRIBUTING.md\` — Changesets flow + dev setup + testing
- \`SECURITY.md\` — vuln disclosure contact
- \`CODE_OF_CONDUCT.md\` — Contributor Covenant 2.1
- \`.github/PULL_REQUEST_TEMPLATE.md\`
- \`.github/ISSUE_TEMPLATE/\` (bug + feature)
- Verify \`.github/dependabot.yml\` covers npm + github-actions

## Acceptance

- [ ] All files exist in the repo
- [ ] README has the embedded-vs-external positioning section
- [ ] Issue templates render correctly when filing a new issue

### LINKED ISSUE #4: Wire Changesets in single-package mode + install changeset-bot
## Context

Acceptance criterion in agent-labs-dev/nebula-desktop#153.

## Scope

- \`pnpm add -D @changesets/cli\`
- \`pnpm changeset init\`
- Configure \`.changeset/config.json\` for single-package mode (\`baseBranch: main\`, \`access: public\`)
- Install the \`changeset-bot\` GitHub app on \`agent-labs-dev/electron-mcp\` — done as part of agent-labs-dev/nebula-desktop#150, verify it's still installed
- Document the changeset workflow in CONTRIBUTING.md (issue #3)

## Acceptance

- [ ] \`.changeset/config.json\` exists and is valid
- [ ] \`changeset-bot\` comments on PRs reminding contributors to add a changeset
- [ ] CONTRIBUTING.md (issue #3) explains when/how to add changesets

### LINKED ISSUE #6: GitHub Actions CI workflow: typecheck + lint + smoke test
## Context

Acceptance criterion in agent-labs-dev/nebula-desktop#153.

## Scope

- \`.github/workflows/ci.yml\`:
  - Triggered on PRs against \`main\` and pushes to \`main\`
  - Runs on Ubuntu (and optionally macOS for the smoke test, since Playwright \`_electron\` exercises the platform binary)
  - Steps: setup-node, setup-pnpm, install, \`pnpm typecheck\`, \`pnpm lint\` (once a linter is wired — separate issue), \`pnpm test\` (smoke + unit when issue #2 lands)
- The smoke test needs a virtual display on Linux runners — wire \`xvfb-run\` or use \`@playwright/test\`'s \`xvfb\` integration
- Cache pnpm + Playwright browsers + Electron binaries to keep CI fast

## Acceptance

- [ ] CI runs on every PR
- [ ] Typecheck failures block merge
- [ ] Smoke test failures block merge
- [ ] Total CI runtime < 5 min for the cached-deps path

### LINKED ISSUE #7: GitHub Actions release workflow: publish via Changesets
## Context

Acceptance criterion in agent-labs-dev/nebula-desktop#153. \`NPM_TOKEN\` was added to repo secrets as part of agent-labs-dev/nebula-desktop#150.

Depends on #5 (Changesets config).

## Scope

- \`.github/workflows/release.yml\` — uses \`changesets/action\` to:
  - Open a \"Version Packages\" PR consolidating pending changesets when commits land on \`main\`
  - On merge of that PR, run \`pnpm publish --access public\` via \`NPM_TOKEN\`
- Tag the release in git, attach a generated GitHub release with changelog

## Acceptance

- [ ] Merging a PR with a changeset opens (or updates) a \"Version Packages\" PR
- [ ] Merging the \"Version Packages\" PR publishes to npm and creates a GitHub release
- [ ] Failed publishes don't leave the repo in a partially-released state

### LINKED ISSUE #8: Add docs/repo-setup.md template for future agent-labs-dev OSS repos
## Context

Acceptance criterion in agent-labs-dev/nebula-desktop#153. Captures the per-repo settings (Dependabot, secret scanning, branch protection, CODEOWNERS, NPM_TOKEN, changeset-bot install) so future \`agent-labs-dev\` OSS repos can be bootstrapped consistently.

## Scope

- \`docs/repo-setup.md\` — checklist + screenshots / gh CLI commands
- Should mirror what was done for this repo as part of agent-labs-dev/nebula-desktop#150

## Acceptance

- [ ] \`docs/repo-setup.md\` exists
- [ ] Following the doc end-to-end produces a repo with the same settings as \`agent-labs-dev/electron-mcp\`

### PR DESCRIPTION
**Pull request by jnsdls**
## Summary

- Extract the embedded Electron MCP server into the standalone `@nebula-agents/electron-mcp` package with explicit lifecycle, generic surface keys, bundled renderer automation tools, and public type exports.
- Add package infrastructure: pnpm lockfile, tsdown build, TypeScript, Biome, Vitest, Changesets, README, contribution/security docs, repo setup notes, issue templates, CI, and release workflow.
- Add unit coverage ported from nebula-desktop plus a Playwright `_electron` smoke fixture for `list_surfaces`, `evaluate`, and `screenshot`.

## Verification

- [x] `pnpm check`
- [x] `pnpm test:electron` if Electron/CDP behavior changed

Notes: `pnpm test:electron` currently skips on macOS because Playwright `_electron` passes `--remote-debugging-port=0`, which Electron 41 rejects on Darwin. The smoke test remains enabled for Linux CI.

## Closes

- Closes #2 (unit tests ported from nebula-desktop)
- Closes #3 (repo docs scaffolding: README, CONTRIBUTING, SECURITY, CODE_OF_CONDUCT, PR/issue templates)
- Closes #4 (Changesets wired in single-package mode)
- Closes #6 (CI workflow: typecheck + lint + smoke test)
- Closes #7 (release workflow via Changesets)
- Closes #8 (docs/repo-setup.md)

## Summary by CodeRabbit

* **New Features**
  * Initial prerelease of the Electron MCP package with embedded MCP server and user tools (screenshot, click, type, form-fill, accessibility snapshot, DOM queries, show/hide/focus, reload, hover, press-key, etc.) and a smoke-test entrypoint.

* **Documentation**
  * Expanded README, CONTRIBUTING, CODE_OF_CONDUCT, SECURITY, and repo-setup docs with quickstart, API, security reporting, and maintenance notes.

* **API**
  * Public config extended with optional path, serverName, serverVersion, and logger; new logger type exported.

* **Testing & CI/CD**
  * Added unit and Electron smoke tests, Playwright/Vitest configs, CI and release workflows, and Changesets.

* **Chores**
  * Packaging and build/tooling updates and lint/formatter config added.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+jobs:
+  check:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v6
+      - uses: pnpm/action-setup@v4
+      - uses: actions/setup-node@v6
+        with:
+          node-version: 24
+          cache: pnpm
```
[the remark this answers is not in the document]
Comment (jnsdls) [reply .github/workflows/ci.yml]: Pinned `actions/checkout@v6`, `pnpm/action-setup@v4`, `actions/setup-node@v6` (and `changesets/action@v1` in `release.yml`) to full commit SHAs in 391daab. Dependabot's `github-actions` ecosystem (already configured in `.github/dependabot.yml`) will keep them current.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+pnpm check
+```
+
+## Development
+
+- Source lives in `src/`.
+- Unit tests are colocated as `*.test.ts`.
+- Electron smoke coverage lives under `test/electron/`.
+- Run `pnpm test` for unit tests.
+- Run `pnpm test:electron` for the Electron smoke test.
```
[the remark this answers is not in the document]
Comment (jnsdls) [reply CONTRIBUTING.md:25]: Clarified in 391daab: `test/electron/smoke.spec.ts` is the canonical Electron smoke (driven by `pnpm test:electron`); `tests/smoke.electron.test.ts` is an unwired exploratory fixture, kept only for local experimentation and may be removed in a follow-up.

File: src/cdp.ts [hunk 3]
```diff
@@ -48,32 +47,15 @@ export function getOrAttachSession(
   const wc = win.webContents;
 
   const existing = attached.get(surface);
-  if (existing && existing.win === win && wc.debugger.isAttached()) {
+  if (existing && !existing.win.isDestroyed() && wc.debugger.isAttached()) {
     return buildSession(existing.surface, wc);
```
[the remark this answers is not in the document]
Comment (jnsdls) [reply src/cdp.ts:56]: Added `existing.win.webContents === wc` guard before reusing the cached session in 391daab.

File: src/cdp.ts [hunk 4]
```diff
@@ -89,19 +71,11 @@ export function getOrAttachSession(
...
-    wc.debugger.off("detach", onDetach);
     console.warn(
       `[mcp] debugger detached from "${surface}" (reason: ${reason})`,
     );
-    const rec = attached.get(surface);
-    if (rec && rec.win.webContents === wc) {
-      attached.delete(surface);
-    }
+    attached.delete(surface);
   };
```
[the remark this answers is not in the document]
Comment (jnsdls) [reply src/cdp.ts:92]: Both detach paths fixed in 391daab: `onDetach` now self-unregisters via `wc.debugger.off("detach", onDetach)` before any cache mutation and only deletes the entry when `rec?.win.webContents === wc`. `CdpSession.detach()` likewise gates the `teardownListener()` + `attached.delete()` on the same binding-equality check so a stale handle from a prior binding can't evict the entry now serving a different window.

File: src/server.ts [hunk 5]
```diff
@@ -166,10 +153,10 @@ export async function startMcpServer(
 
   const httpServer = createServer(async (req, res) => {
     try {
-      await routeRequest(req, res, sessions, createSession, () => shuttingDown);
+      await routeRequest(req, res, sessions, createSession, path);
```
[the remark this answers is not in the document]
Comment (jnsdls) [reply src/server.ts]: Reintroduced the shutdown latch in 391daab: `shuttingDown` is set inside `stop()` before snapshotting sessions, and `routeRequest` short-circuits with a 503 `Server shutting down` JSON-RPC error so a racing initialize can't mint a post-snapshot transport that hangs `httpServer.close()`.

File: src/surfaces.ts [hunk 6]
```diff
@@ -22,15 +22,7 @@ export function resolveSurface(
...
-  // plain `surfaces[surface]` lookup would return `Object.prototype`
-  // for `"__proto__"`, which is not a `BrowserWindow` and would throw
-  // a confusing `isDestroyed is not a function` instead of the
-  // intended `SurfaceNotFoundError`.
-  const surfaces = getSurfaces();
-  const win = Object.prototype.hasOwnProperty.call(surfaces, surface)
-    ? surfaces[surface]
-    : null;
+  const win = getSurfaces()[surface];
   if (!win || win.isDestroyed()) {
```
[the remark this answers is not in the document]
Comment (jnsdls) [reply src/surfaces.ts]: Added an own-property guard via `Object.prototype.hasOwnProperty.call(...)` in 391daab so keys like `__proto__` or `constructor` no longer walk up the prototype chain — they now surface `SurfaceNotFoundError` instead of throwing a raw `TypeError` from `isDestroyed()` on a function value.

File: src/testing.ts [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+        const response = cdpResponses[method];
+        if (typeof response === "function") {
+          return await (
+            response as (p: Record<string, unknown> | undefined) => unknown
+          )(params);
+        }
+        return response ?? {};
+      },
+      on: vi.fn(),
+      off: vi.fn(),
```
[the remark this answers is not in the document]
Comment (jnsdls) [reply src/testing.ts]: Reworked the fake debugger in 391daab: `sendCommand` now throws when not attached so tools that mint a stale handle and keep using it after teardown surface as a test failure, and `debugger.on/off` actually track listeners (instead of being bare `vi.fn()` no-ops). Added `emitDebugger` + `debuggerListenerCount` escape hatches on the fake so future tests can fire a real `detach` event and assert listener cleanup in `CdpSession`.

File: src/tools/ax-snapshot.test.ts [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    }));
+    const tool = getTool(server, "get_ax_snapshot");
+    const result = (await tool.handler({ surface: "ax_main" })) as {
+      structuredContent: { nodes: Array<{ id: string; children: string[] }> };
+    };
+    // Ignored child filtered out + child reference scrubbed.
+    expect(result.structuredContent.nodes).toHaveLength(1);
+    expect(result.structuredContent.nodes[0]).toMatchObject({
+      id: "1",
+      children: [],
```
[the remark this answers is not in the document]
Comment (jnsdls) [reply src/tools/ax-snapshot.test.ts:58]: Push back: this finding has the direction inverted. The CDP-shaped *input* nodes use `nodeId` (see the mocked `Accessibility.getFullAXTree` payload in the same test, line 36, and `AxNode.nodeId` in [`ax-snapshot.ts:34`](https://github.com/agent-labs-dev/electron-mcp/blob/391daab/src/tools/ax-snapshot.ts#L34)), but `formatTree` deliberately renames them to `id` in the *output* (`id: n.nodeId` at [`ax-snapshot.ts:161`](https://github.com/agent-labs-dev/electron-mcp/blob/391daab/src/tools/ax-snapshot.ts#L161)). The test asserting `{ id: "1", children: [] }` matches the formatter contract — switching it to `nodeId` would break the test against the actual runtime shape. Tests pass against this assertion (`pnpm check` is green on 391daab).

File: src/tools/list-surfaces.test.ts [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+import { registerListSurfaces } from "./list-surfaces";
+
+describe("list_surfaces", () => {
+  it("registers a tool with no input schema", () => {
+    const server = createFakeMcpServer();
+    registerListSurfaces(server.asMcpServer, () => ({}));
+    const tool = getTool(server, "list_surfaces");
+    expect(tool.config.inputSchema).toBeUndefined();
+    expect(tool.config.title).toBe("List surfaces");
+  });
```
[the remark this answers is not in the document]
Comment (jnsdls) [reply src/tools/list-surfaces.test.ts:16]: Push back: the "every tool's inputSchema must accept `surface: string`" rule applies to per-surface tools (`screenshot`, `evaluate`, `click`, etc.), not to the meta-enumerator. `list_surfaces` is the discovery primitive — its job is to *return* the set of surfaces, so requiring a surface arg would be circular. The implementation in [`list-surfaces.ts`](https://github.com/agent-labs-dev/electron-mcp/blob/391daab/src/tools/list-surfaces.ts) intentionally registers without an `inputSchema` and the handler ignores its input; the test codifies that contract correctly. Open to documenting this carve-out in the coding guidelines as a follow-up if it would be helpful.

**Comment by jnsdls** [comment]
@coderabbitai — also addressed the items from the review body in 391daab:

- **Outside-diff comment on `src/server.ts:247-264`**: `req.socket?.destroy()` is now passed as the `res.end` callback so the 413 JSON-RPC body flushes before the connection tears down (no more ECONNRESET on the client).
- **Nitpick on `src/tools/evaluate.ts:42-57`**: dropped the unused `unserializableValue` field from `RuntimeEvaluateResult` since the formatter falls back to `description`/`type`.
- **Nitpick on `package.json`**: removed the duplicate `async-mutex` from `devDependencies`; it's a runtime import so `dependencies` is the authoritative entry.

Two findings I pushed back on (replies on the inline threads):
- `src/tools/ax-snapshot.test.ts` (`id` vs `nodeId`) — the formatter intentionally renames `nodeId` → `id`, so the existing assertion is correct.
- `src/tools/list-surfaces.test.ts` — `list_surfaces` is the meta-enumerator and intentionally registers without an `inputSchema`; the rule "every tool requires `surface`" applies to per-surface tools.

File: src/testing.ts [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+  const once = (event: string, listener: Listener) => {
+    const bucket = onceListeners.get(event) ?? new Set();
+    bucket.add(listener);
+    onceListeners.set(event, bucket);
+  };
+  const emit = (event: string, ...args: unknown[]) => {
+    for (const l of listeners.get(event) ?? []) l(...args);
+    for (const l of onceListeners.get(event) ?? []) l(...args);
+    onceListeners.get(event)?.clear();
+  };
```
[the remark this answers is not in the document]
Comment (jnsdls) [reply src/testing.ts:175]: Fixed in 6d5e105: snapshot both listener buckets first, `onceListeners.delete(event)` before invocation, then iterate the snapshot. A handler that calls `emit()` recursively or adds a new listener mid-flight no longer sees the already-firing batch, and stale once-listeners can't refire.
